test: fix MiniWallet script-path spend (missing parity bit in leaf version)

theStack opened the pull request:
This PR fixes a dormant bug in MiniWallet that exists since support for P2TR was initially added in #23371 (see commit 041abfebe49ae5e3e882c00cc5caea1365a27a49).
    
In the course of spending the output, the leaf version byte of the control block in the witness stack doesn't set the parity bit, i.e. we were so far just lucky that the used combinations of relevant data (internal pubkey, leaf script / version) didn't result in a tweaked pubkey with odd y-parity. If that was the case, we'd get the following validation error:

`mandatory-script-verify-flag-failed (Witness program hash mismatch) (-26)`
    
Since MiniWallets can now optionally be tagged (#29939), resulting in different internal pubkeys, the issue is more prevalent now. Fix it by passing the parity bit, as specified in BIP341.

Can be tested with the following patch (fails on master, succeeds on PR):
```diff
diff --git a/test/functional/test_framework/mempool_util.py b/test/functional/test_framework/mempool_util.py
index 148cc935ed..7ebe858681 100644
--- a/test/functional/test_framework/mempool_util.py
+++ b/test/functional/test_framework/mempool_util.py
@@ -42,7 +42,7 @@ def fill_mempool(test_framework, node):
     # Generate UTXOs to flood the mempool
     # 1 to create a tx initially that will be evicted from the mempool later
     # 75 transactions each with a fee rate higher than the previous one
-    ephemeral_miniwallet = MiniWallet(node, tag_name="fill_mempool_ephemeral_wallet")
+    ephemeral_miniwallet = MiniWallet(node, tag_name="fill_mempool_ephemeral_wallet3")
     test_framework.generate(ephemeral_miniwallet, 1 + num_of_batches * tx_batch_size)
 
     # Mine enough blocks so that the UTXOs are allowed to be spent
```

In addition to that, another bug is fixed where the internal key derivation failed, as not every pseudorandom hash results in a valid x-only pubkey. Fix this by treating the hash result as private key and calculate the x-only public key out of that, to be used then as internal key.

Fixes #30528.

theStack committed:
test: refactor: return TaprootInfo from P2TR address creation routine

Rather than only returning the internal key from the P2TR anyone-can-spend
address creation routine, provide the whole TaprootInfo object, which in turn
contains a dictionary of TaprootLeafInfo object for named leaves.

This data is used in MiniWallet for the default ADDRESS_OP_TRUE mode, in order
to deduplicate the witness script and leaf version of the control block.

theStack committed:
test: fix MiniWallet script-path spend (missing parity bit in leaf version)

This commit fixes a dormant bug in MiniWallet that exists since
support for P2TR was initially added in #23371 (see commit
041abfebe49ae5e3e882c00cc5caea1365a27a49).

In the course of spending the output, the leaf version byte of the
control block in the witness stack doesn't set the parity bit, i.e.
we were so far just lucky that the used combinations of relevant
data (internal pubkey, leaf script / version) didn't result in a
tweaked pubkey with odd y-parity. If that was the case, we'd get the
following validation error:

`mandatory-script-verify-flag-failed (Witness program hash mismatch) (-26)`

Since MiniWallets can now optionally be tagged (#29939), resulting
in different internal pubkeys, the issue is more prevalent now.
Fix it by passing the parity bit, as specified in BIP341.

rkrux commented on test/functional/test_framework/wallet.py:
```diff
@@ -187,7 +186,11 @@ def sign_tx(self, tx, fixed_length=True):
             tx.wit.vtxinwit = [CTxInWitness()] * len(tx.vin)
             for i in tx.wit.vtxinwit:
-                i.scriptWitness.stack = [CScript([OP_TRUE]), bytes([LEAF_VERSION_TAPSCRIPT]) + self._internal_key]
+                leaf_info = self._taproot_info.leaves["only-path"]
```
I feel we should tie this `only-path` here to `taproot_construct` call in `address.py`, either by passing it as a parameter or creating it as a constant. Otherwise, it makes you wonder here at this line where did `only-path` come from.

rkrux replied:
Yes, this is also good.

rkrux commented on test/functional/test_framework/wallet.py:
```diff
@@ -187,7 +186,11 @@ def sign_tx(self, tx, fixed_length=True):
             for i in tx.wit.vtxinwit:
-                i.scriptWitness.stack = [CScript([OP_TRUE]), bytes([LEAF_VERSION_TAPSCRIPT]) + self._internal_key]
+                leaf_info = self._taproot_info.leaves["only-path"]
+                i.scriptWitness.stack = [
+                    leaf_info.script,
+                    bytes([leaf_info.version | self._taproot_info.negflag]) + self._taproot_info.internal_pubkey,
+                ]
```
Would love to see the corresponding cpp reference code, share if you know?

rkrux replied:
Great, thanks for sharing, I'm digging into this!

rkrux reviewed (approved):
tACK [e4abac2](https://github.com/bitcoin/bitcoin/pull/30076/commits/e4abac2fce4af98108d899bc0480d066e5da7754)

Reviewing #29939 previously made it easy for me to review this one! 

Make successful and all functional tests pass on this branch. Using the shared patch, I can see these 2 tests fail in master branch.

```
feature_versionbits_warning.py                         | ✖ Failed  | 3 s
wallet_fundrawtransaction.py --descriptors             | ✖ Failed  | 17 s

ALL                                                    | ✖ Failed  | 2752 s (accumulated)
```

rkrux commented:
@theStack Thanks for the quick update on this, reACK [57eb590](https://github.com/bitcoin/bitcoin/pull/30076/commits/57eb59024b4d5b776ddb08736a81585c2ea0e734)

theStack committed:
test: fix MiniWallet internal key derivation for tagged instances

Not every pseudorandom hash result is a valid x-only public key,
so the pubkey tweaking in the course of creating the output public
key would fail about every second time.

Fix this by treating the hash result as private key and calculate
the x-only public key out of that, to be used then as internal key.

theStack committed:
test: add functional test for tagged MiniWallet instances

glozow reviewed (commented):
ACK e4b0dabb2115dc74e9c5794ddca3822cd8301c72

Not an expert but: Checked that `test_wallet_tagging` fails without the fix on "mandatory-script-verify-flag-failed (Witness program hash mismatch) (-26)". Reviewed the test by editing the parity check in `VerifyTaprootCommitment` to make it pass on master and examining `bytes(version + negflag)` while running.

rkrux reviewed (approved):
reACK [e4b0dab](https://github.com/bitcoin/bitcoin/pull/30076/commits/e4b0dabb2115dc74e9c5794ddca3822cd8301c72)

Applied the shared patch again and the following tests fail on master with the mentioned error. Though I've not been able to go through the corresponding control block C++ code yet :(

```
raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: mandatory-script-verify-flag-failed (Witness program hash mismatch) (-26)
```

```
mempool_limit.py                                         | ✖ Failed  | 1 s
p2p_1p1c_network.py                                      | ✖ Failed  | 5 s
p2p_opportunistic_1p1c.py                                | ✖ Failed  | 1 s
p2p_tx_download.py                                       | ✖ Failed  | 31 s
rpc_packages.py                                          | ✖ Failed  | 3 s
```

rkrux commented on test/functional/feature_framework_miniwallet.py:
```diff
@@ -31,6 +34,20 @@ def test_tx_padding(self):
+        self.log.info(f"Test tagged wallet instances...")
+        node = self.nodes[0]
+        untagged_wallet = self.wallets[0][1]
+        for i in range(10):
```
Is the sole reason for creating 10 tagged wallets to be thorough? Or is there any other reason as well?

rkrux commented:
Got it, nice.

On Tue, Jun 18, 2024, 06:43 Sebastian Falbesoner ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/functional/feature_framework_miniwallet.py
> <https://github.com/bitcoin/bitcoin/pull/30076#discussion_r1643600316>:
>
> > @@ -31,6 +34,20 @@ def test_tx_padding(self):
>                  assert_greater_than_or_equal(tx.get_weight(), target_weight)
>                  assert_greater_than_or_equal(target_weight + 3, tx.get_weight())
>
> +    def test_wallet_tagging(self):
> +        """Verify that tagged wallet instances are able to send funds."""
> +        self.log.info(f"Test tagged wallet instances...")
> +        node = self.nodes[0]
> +        untagged_wallet = self.wallets[0][1]
> +        for i in range(10):
>
> Is the sole reason for creating 10 tagged wallets to be thorough? Or is
> there any other reason as well?
>
> Yes it's thoroughness, no other reason. Since the fixed bugs only occur
> with a certain probability, creating several random tagged wallets in a row
> increase the likelihood of catching the bugs in a single test run (nothing
> special about the number 10 though).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/30076#discussion_r1643600316>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNPILUIX2ZRNELG5ECQDSLZH6CT3AVCNFSM6AAAAABHPTP73OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCMRUGE2TAMBSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hodlinator reviewed (approved):
ACK e4b0dabb2115dc74e9c5794ddca3822cd8301c72

Confirmed to fix #30528 which I discovered independently. Rebased my test commit from there on top of this PR and verified that exceptions are no longer raised for the 2 cases I found.

Re-ran **feature_framework_miniwallet.py** without my changes on top and verified it passed. Also ran with `--randomseed <X>` to verify that tag names are generated deterministically.

Passed `test_runner.py` (other tests are using `MiniWallet` indirectly, and **mempool_util.py** specifically was already using a `tag_name` value that still works).

Not an expert on Taproot so haven't fully understood the context around the fixes, but they look small & safe and only touch functional test Python code.

Sorry again for not finding this PR before creating the issue. Eager to use this fix to fearlessly create multiple tagged wallets in upcoming tests. :)
